nearswap/src/twap.rs

```diff
@@ -0,0 +1,601 @@
+                1, 1,
+                max_length
+            );
+        }
+
+        assert!(twap.observations.len() == 10, "Mismatch");
+
+        // next observation should be written on 0th Index
+        let mut last_timestamp = env::block_timestamp() + 10;
+        current_idx = twap.write(
+            last_timestamp,
+            1, 1,
+            max_length
+        );
+
+        assert!(twap.observations.len() == 10, "Mismatch");
+        assert!(twap.observations.get(0).unwrap().block_timestamp == last_timestamp, "Mismatch");
+        assert!(twap.observations.get(0).unwrap().num_of_observations == 11);
+
+        // next observation should be written on 1st Index
+        last_timestamp = env::block_timestamp() + 11;
+        current_idx = twap.write(
+            last_timestamp,
+            1, 1,
+            max_length
+        );
+
+        env_log!("as {}", twap.observations.len());
+        assert!(twap.observations.len() == 10, "Mismatch");
+        assert!(current_idx == 1, "current index mismatch");
+
+        assert!(twap.observations.get(1).unwrap().block_timestamp == last_timestamp, "Mismatch");
+        assert!(twap.observations.get(1).unwrap().num_of_observations == 12);
+    }
+
+    #[test]
+    fn simple_binary_search_works() {
+        init_blockchain();
+
+        let twap: Twap = get_twap(env::block_timestamp());
+        let max_length = 10;
+
+        // current observation timestamp array [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+        let mut returned_index = twap.binary_search(
+            max_length,
+            5,
+        );
+
+        assert!(returned_index == 4, "Wrong Index");
+
+        returned_index = twap.binary_search(
+            max_length,
+            0,
+        );
+
+        assert!(returned_index == 0, "Wrong Index");
+
+        returned_index = twap.binary_search(
+            max_length,
+            10,
+        );
+
+        assert!(returned_index == 9, "Wrong Index");
+    }
+
+    #[test]
+    #[should_panic(expected = "Observation after this timestamp doesn't exist")]
+    fn binary_edge_case_works() {
+        init_blockchain();
+
+        let twap: Twap = get_twap(env::block_timestamp());
+        let max_length = 10;
+
+        // current observation timestamp array [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+        twap.binary_search(
+            max_length,
+            122,
+        );
+    }
+
+    #[test]
+    fn pivoted_binary_search_works() {
+        init_blockchain();
+
+        let mut twap: Twap = get_twap(env::block_timestamp());
+        let max_length = 10;
+
+        // current array [1, 2, 3, 4, 5, 6, 8, 9, 10]
+        // add more value (that should overwrite last updated value)
+        let mut current_idx = twap.write(
+            13,
+            10, 10,
+            max_length
+        );
+
+        let mut result_index = twap.binary_search(
+            max_length,
+            11,
+        );
+        env_log!("SSS {} {}", result_index, current_idx);
```

Comment:
why not using `println!`? in this test let's don't use `env::` functions.

Reply:
ok